core/src/main/java/software/amazon/awssdk/async/SingleByteArrayAsyncRequestProvider.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.async;
+
+import java.nio.ByteBuffer;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.utils.builder.SdkBuilder;
+
+class SingleByteArrayAsyncRequestProvider implements AsyncRequestProvider {
+
+    private final byte[] bytes;
+
+    private SingleByteArrayAsyncRequestProvider(DefaultBuilder builder) {
+        this.bytes = builder.bytes.clone();
+    }
+
+    @Override
+    public long contentLength() {
+        return bytes.length;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super ByteBuffer> subscriber) {
+        subscriber.onSubscribe(
+            new Subscription() {
+                @Override
+                public void request(long n) {
+                    if (n > 0) {
+                        subscriber.onNext(ByteBuffer.wrap(bytes));
+                        subscriber.onComplete();
+                    }
+                }
+
+                @Override
+                public void cancel() {
+                }
+            }
+        );
+    }
+
+    public static Builder builder() {
+        return new DefaultBuilder();
+    }
+
+    public interface Builder extends SdkBuilder<Builder, SingleByteArrayAsyncRequestProvider> {
+        Builder bytes(byte[] bytes);
+    }
+
+    public static class DefaultBuilder implements Builder {
```

Comment:
For internal stuff I think we can skip the builder.

Reply:
Gotcha, I made that public since there were some options worth exposing. If we don't expose the class I'd stick with a constructor for now.